irods/test/data_obj_test.py

```diff
@@ -654,6 +675,168 @@ def test_multiple_reads(self):
+            error = cls.WrongUserType('Must be an iRODS admin to run tests in class {0.__name__}'.format(cls))
+            raise error
+        cls.logins = iRODSUserLogins(adm)
+        cls.logins.create_user(RODSUSER,'abc123')
+
+    @classmethod
+    def tearDownClass(cls):
+        # TODO: Skipping this will result in an interpreter seg fault for Py3.6 but not 3.11; why?
+        del cls.logins
+
+    def _data_object_and_associated_ticket(self, 
+                                           data_name='',             # A random name will be generated by default.
+                                           content=None,             # Content to use if creating data object internally.
+                                           auto_delete_data = True,  # Whether to delete up the data object in finalization.
+                                           ticket_access=''          # No ticket is generated unless string is nonzero length.
+                                           ):
+        ticket = [None]
+        data = [None]
+        if '/' in data_name:
+            data_path = data_name
+        else:
+            if not data_name:
+                data_name = helpers.unique_name(helpers.my_function_name(), datetime.now())
+            data_path = helpers.home_collection(self.sess)+"/"+data_name
+        def create_data(open_options):
+            if content is not None and not self.sess.data_objects.exists(data_path):
+                with self.sess.data_objects.open(data_path,"w",**dict(open_options)) as f:
+                    f.write(content)
+            data[0] = data_path
+        try:
+            if ticket_access:
+                user_session = self.logins.session_for_user(RODSUSER)
+                def initialize(open_options=()):
+                    create_data(open_options)
+                    # Activate ticket for test session:
+                    ticket_logical_path = data_path if self.sess.data_objects.exists(data_path) else irods_dirname(data_path)
+                    ticket[0] = Ticket(self.sess).issue(ticket_access, ticket_logical_path)
+                    Ticket(user_session, ticket[0].string).supply()
+            else:
+                user_session = self.sess
+                initialize = lambda open_options=(): \
+                                 create_data(open_options)
+            # Can be called early to clear resources of data objects before their deletion
+            def finalize():
+                if auto_delete_data and data[0]:
+                    self.sess.data_objects.unlink(data[0], force=True)
+                    data[0]=None
+            yield { 'path':data_path, 'session':user_session, 'finalize': finalize, 'initialize':initialize, 'ticket_access': ticket_access }
+        finally:
+            if ticket[0]:
+                ticket[0].delete()
+            finalize()
+
+    data_object_and_associated_ticket = contextlib.contextmanager(_data_object_and_associated_ticket)
```

Comment:
Can you explain the reasoning behind the indirection here? Why not just have the tests call `_data_object_and_associated_ticket` directly?

Reply:
Shortening `data_object_and_associated_ticket` to `{XYZ}` for more succinct reference, I'll summarize below what's being done here.

We're making two forms of the context-manager, `_{XYZ}` and `{XYZ}`.
   - `_{XYZ}` is the raw form, essentially a generator function which provides the with-block's `as`-value via a `yield` plus some `__exit__`-style `with`-block cleanup.  This raw form can be manipulated like a co-routine to "emulate"  with-block style context in such places as `do_test_redirect_in_data_object_put_and_get__issue_452` where the with-form cannot be directly used because it's conditionally called (and sadly, you can't do an IF-WITH in Python). So the actual with-block is emulated between the `generator =... ` and the `del generator` statements.
   - `{XYZ}` is the cooked form, transformed into an actual `with`-block compatible object via Python's `contextlib.contextmanager` convenience function.  We then use that cooked form in an actual  with-block in another test method, `test_redirect_in_data_object_put_and_get_without_tickets__issue_452`

Reply:
Seems reasonable to me. Thanks for the explanation